### PR TITLE
Ni converting functional constraints class

### PR DESCRIPTION
**Pull request by niharika2999**
PR for converting functional constraints to a class.

### CODE REVIEW

File: pysensors/utils/__init__.py [hunk 1]
```diff
@@ -3,10 +3,14 @@
...
-from ._constraints import get_coordinates_from_indices
-from ._constraints import get_indices_from_coordinates
+from ._constraints import BaseConstraint
+from ._constraints import Circle
+from ._constraints import Line
+from ._constraints import UserDefinedConstraints
+# from ._constraints import constraints_eval
+# from ._constraints import functional_constraints
+# from ._constraints import get_coordinates_from_indices
+# from ._constraints import get_indices_from_coordinates
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/__init__.py]: Should these lines be deleted, or might they be useful in the future?  Also, `constraints_eval` and others are still listed in the `__all__`, is that correct?

File: pysensors/utils/_constraints.py [hunk 2]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
-    dirName = os.path.dirname(functionHandler)
-    sys.path.insert(0,os.path.expanduser(dirName))
-    module = __import__(functionName)
-    func = getattr(module, functionName)
-    g = func(xLoc, yLoc,**kwargs)
-    return g
-
+class BaseConstraint(object):
+    '''
+    --To be filled---
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: This needs to be filled out

File: pysensors/utils/_constraints.py [hunk 3]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+        """
+        if 'data' in kwgs.keys():
+            self.data = kwgs['data']
+        else:
+            raise Exception('Must provide data as **kwgs')
+        if isinstance(self.data,pd.DataFrame):
+            if 'X_axis' in kwgs.keys():
+                self.X_axis = kwgs['X_axis']
+            else:
+                raise Exception('Must provide Y_axis as **kwgs as your data is a dataframe')
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Y_axis should be X_axis

File: pysensors/utils/_constraints.py [hunk 4]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+            if 'Field' in kwgs.keys():
+                self.Field = kwgs['Field']
+            else:
+                raise Exception('Must provide Field as **kwgs as your data is a dataframe')
+        
+    
+    def get_constraint_indices(self,all_sensors,info):
+        '''
+        To be Filled
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Please expand this comment

File: pysensors/utils/_constraints.py [hunk 5]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+        g = np.zeros(len(x),dtype = float)
+        for i in range(len(x)):
+             g[i] = self.constraint_function(x[i], y[i])
+        G_const = constraints_eval([g],all_sensors,data = info)
+        idx_const, rank = get_functionalConstraind_sensors_indices(all_sensors,G_const[:,0])
+        return idx_const,rank
+    
+    def draw_constraint(self):
+        '''
+        To be Filled
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Please expand this comment.

File: pysensors/utils/_constraints.py [hunk 6]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+            image : if the data is represented in the fprm of an image
+            scatter: if the data can be represented with a scatter plot
+            contour_map: if the data can be represented in the form of a contour map
+        ** kwargs : Required for data in the form of a dataframe for the scatter and contour plots
+            X_axis : string,
+                Name of the column in dataframe to be plotted on the X axis.
+            Y-axis : string,
+                Name of the column in dataframe to be plotted on the Y axis.
+            Field : string,
+                Name of the column in dataframe to be plotted as a contour map.
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: It looks like kwargs are not used in this function. Does the comment need updating?

File: pysensors/utils/_constraints.py [hunk 7]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+        if isinstance(self.data,pd.DataFrame):
+            if 'X_axis' in kwgs.keys():
+                self.X_axis = kwgs['X_axis']
+            else:
+                raise Exception('Must provide Y_axis as **kwgs as your data is a dataframe')
+            if 'Y_axis' in kwgs.keys():
+                self.Y_axis = kwgs['Y_axis']
+            else:
+                raise Exception('Must provide Y_axis as **kwgs as your data is a dataframe')
+            if 'Field' in kwgs.keys():
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Hm, should Field be optional? It only seems to be used on contour plots.

File: pysensors/utils/_constraints.py [hunk 8]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+        elif plot_type == 'contour_map': 
+            fig , ax = plt.subplots()
+            y_vals = self.data[self.Y_axis]
+            x_vals = self.data[self.X_axis]
+            ax.scatter(x_vals, y_vals, c = self.data[self.Field], cmap = plt.cm.coolwarm, s = 1)
+        self.draw(ax)
+        
+    def plot_grid(self,all_sensors):
+        '''
+        To be filled
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Please fill out.

File: pysensors/utils/_constraints.py [hunk 9]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+            ax.scatter(x_val, y_val, color = 'blue', marker = '.')
+        elif isinstance(self.data,pd.DataFrame):
+            y_vals = self.data[self.Y_axis]
+            x_vals = self.data[self.X_axis]
+            fig , ax = plt.subplots()
+            ax.scatter(x_vals, y_vals, color = 'blue', marker = '.')
+        
+    def sensors_dataframe(self,sensors):
+        '''
+        To be filled
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Please expand this comment.

File: pysensors/utils/_constraints.py [hunk 10]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+        columns = ['Sensor ID','SensorX','sensorY'] 
+        Sensors_df = pd.DataFrame(data = np.vstack([sensors,xTop,yTop]).T,columns=columns,dtype=float)
+        Sensors_df.head(n_sensors)
+        return Sensors_df
+        
+            
+        
+    def annotate_sensors(self,sensors):
+        '''
+        To be filled
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Please expand this comment.

File: pysensors/utils/_constraints.py [hunk 11]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+        if isinstance(self.data,np.ndarray):
+            xTop = np.mod(sensors,np.sqrt(n_features))
+            yTop = np.floor(sensors/np.sqrt(n_features))
+            data = np.vstack([sensors,xTop,yTop]).T
+            for ind,i in enumerate(range(len(xTop))):
+                plt.annotate(f"{str(ind)}",(xTop[i],yTop[i]),xycoords='data',
+                    xytext=(-20,20), textcoords='offset points',color="r",fontsize=12,
+                    arrowprops=dict(arrowstyle="->", color='black'))
+        elif isinstance(self.data,pd.DataFrame):
+            X,Y = self.data[self.X_axis], self.data[self.Y_axis]
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Are X and Y used? (Could this line be deleted?)

File: pysensors/utils/_constraints.py [hunk 12]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+            y-coordinate of the center of circle
+        radius : float,
+            radius of the circle
+        '''
+        self.center_x = center_x
+        self.center_y = center_y
+        self.radius = radius
+        
+    def draw(self,ax):
+        c = patches.Circle((self.center_x, self.center_y), self.radius, fill = False, color = 'r', lw = 2)
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Please add a docstring.

File: pysensors/utils/_constraints.py [hunk 13]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+        ax.add_patch(c)
+        ax.autoscale_view()
+        
+        
+    def constraint_function(self,x_all_unc, y_all_unc):
+        '''
+        To be Filled
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Please expand this comment.

File: pysensors/utils/_constraints.py [hunk 14]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+        self.center_y = center_y
+        self.radius = radius
+        
+    def draw(self,ax):
+        c = patches.Circle((self.center_x, self.center_y), self.radius, fill = False, color = 'r', lw = 2)
+        ax.add_patch(c)
+        ax.autoscale_view()
+        
+        
+    def constraint_function(self,x_all_unc, y_all_unc):
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Could x_all_unc, y_all_unc be changed to just x,y? (Ditto for other uses in this file.)

File: pysensors/utils/_constraints.py [hunk 15]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+        '''
+        To be Filled
+        '''
+        return ((x_all_unc-self.center_x)**2 + (y_all_unc-self.center_y)**2) - self.radius**2
+        
+           
+class Line(BaseConstraint):
+    '''
+    To be filled
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Please expand this comment.

File: pysensors/utils/_constraints.py [hunk 16]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+            y-coordinate of the other end-point of the line
+        '''
+        self.x1 = x1
+        self.x2 = x2
+        self.y1 = y1
+        self.y2 = y2
+        
+    def draw(self,ax):
+        '''
+        To be Filled
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Please fill out comment.

File: pysensors/utils/_constraints.py [hunk 17]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+        '''
+        ax.plot([self.x1,self.x2],[self.y1,self.y2],'-r')
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Hm, I don't think this gives an indication which side is the constrained region.

File: pysensors/utils/_constraints.py [hunk 18]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+            
+    def constraint_function(self,x_all_unc, y_all_unc):
+        '''
+        To be Filled
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Please expand comment.

File: pysensors/utils/_constraints.py [hunk 19]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+        h : float,
+            x-coordinate of the vertex of the parabola we want to be constrained
+        k : float,
+            y-coordinate of the vertex of the parabola we want to be constrained
+        a : float,
+            x-coordinate of the focus of the parabola
+        '''
+        self.h = h
+        self.k = k
+        self.a = a
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py:867]: This is missing the constraint_function

File: pysensors/utils/_constraints.py [hunk 20]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+    General class for dealing with any form of user defined constraints.
+    Plotting, computing constraints functionalities included. 
+    '''
+    def __init__(self,all_sensors,info,const_path, **kwgs):
+        super().__init__(**kwgs)
+        '''
+        Attributes
+        ----------
+        const_path : 'string',
+            path of the file that constains the user-defined constraint function
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: What does the file need to contain? Please explain in the comment.

File: pysensors/utils/_constraints.py [hunk 21]
```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+        '''
+        self.all_sensors = all_sensors
+        self.info = info
+        self.const_path = const_path
+        
+    def draw(self,ax):
+        '''
+        To be filled
+        '''
+        nConstraints = len([self.const_path])
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: Um, why are we taking the length of a string and using that as the number of constraints?

File: pysensors/utils/_constraints.py [hunk 22]
```diff
@@ -185,10 +490,10 @@ def get_coordinates_from_indices(idx,info):
     if isinstance(info,tuple):
         return np.unravel_index(idx,info,'F')
     elif isinstance(info,pd.DataFrame):
-        x = info.loc[idx,'X (m)']#.values
+        x = info.loc[idx,'X (m)']#.values  ## This will not always be X (m) or Y (m) CHANGE!!!
```
Comment (joshua-cogliati-inl) [inline pysensors/utils/_constraints.py]: `get_coordinates_from_indices` is called from BaseConstraint, so this needs to be fixed, or remove calling it from BaseConstraint

**Review by joshua-cogliati-inl — COMMENTED** [review]
Here are various things that I think probably need to be fixed before merging it.

File: pysensors/utils/__init__.py (hunk 1 above)
Comment (niharika2999) [reply pysensors/utils/__init__.py]: I think we should leave these lines as they are not a part of the class for functional constraints but helper functions. If a user wants to directly use these maybe they should have access. 
What do you think @Jimmy-INL ?

File: pysensors/utils/_constraints.py (hunk 7 above)
Comment (niharika2999) [reply pysensors/utils/_constraints.py]: Yes we could make it optional for places where the user selects image, scatter in the plot_constraints_on_data() and necessary only if contour_map is the type of plot  however I do not know how to implement this. If you can do it that would be great.

File: pysensors/utils/_constraints.py (hunk 11 above)
Comment (niharika2999) [reply pysensors/utils/_constraints.py]: yes this line can be deleted

File: pysensors/utils/_constraints.py (hunk 14 above)
Comment (niharika2999) [reply pysensors/utils/_constraints.py]: Yes they can be changed to just x and y

File: pysensors/utils/_constraints.py (hunk 7 above)
Comment (joshua-cogliati-inl) [reply pysensors/utils/_constraints.py]: Something like:
```python
if 'Field' in kwgs.keys():
    self.Field = kwgs['Field']
else:
    self.Field = None            
```
and then later:
```python
            if self.Field is not None:
               ax.scatter(x_vals, y_vals, c = self.data[self.Field], cmap = plt.cm.coolwarm, s = 1)
            else:
               raise Exception('Must provide Field if using contour_map')
```

File: pysensors/utils/_constraints.py (hunk 17 above)
Comment (niharika2999) [reply pysensors/utils/_constraints.py]: Yes, I still do not know exactly how do we want to handle on which side of the line is that constraint. Maybe we can have a discussion on this with @Jimmy-INL .

File: pysensors/utils/_constraints.py [hunk 23]
```diff
@@ -64,48 +66,578 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+        -----------
+        Annotation of sensor rank near the choosen sensor locations
+        '''
+        n_samples, n_features = self.data.shape
+        n_sensors = len(sensors)
+        if isinstance(self.data,np.ndarray):
+            xTop = np.mod(sensors,np.sqrt(n_features))
+            yTop = np.floor(sensors/np.sqrt(n_features))
+            data = np.vstack([sensors,xTop,yTop]).T
+            for ind,i in enumerate(range(len(xTop))):
```
Comment (Jimmy-INL) [inline pysensors/utils/_constraints.py:552]: the variable data is not used in this function

File: pysensors/utils/_constraints.py [hunk 24]
```diff
@@ -64,48 +66,578 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
...
+                plt.annotate(f"{str(ind)}",(xTop[i],yTop[i]),xycoords='data',
+                    xytext=(-20,20), textcoords='offset points',color="r",fontsize=12,
+                    arrowprops=dict(arrowstyle="->", color='black'))
+        elif isinstance(self.data,pd.DataFrame):
+            xTop, yTop = get_coordinates_from_indices(sensors,self.data,Y_axis = self.Y_axis, X_axis = self.X_axis, Field = self.Field)    #### Annotate not working for dataframe : FIX
+            data = np.vstack([sensors,xTop,yTop]).T
+            for _,i in enumerate(range(len(sensors))):
```
Comment (Jimmy-INL) [inline pysensors/utils/_constraints.py:562]: variable data is not used in this function

File: pysensors/utils/__init__.py (hunk 1 above)
Comment (Jimmy-INL) [reply pysensors/utils/__init__.py]: Niharika, I think @joshua-cogliati-inl meant the lines that are commented out.

File: pysensors/utils/_constraints.py (hunk 2 above)
Comment (Jimmy-INL) [reply pysensors/utils/_constraints.py]: @niharika2999, could you please fill in the docstring?

File: pysensors/utils/_constraints.py (hunk 3 above)
Comment (Jimmy-INL) [reply pysensors/utils/_constraints.py]: @niharika2999 I remember I have changed this before. Now I am questioning if this was on a different PR?

**Pull request merged**